[feat] AOP를 적용한 SSE 기반 알림 기능 구현

JeongUijeong opened the pull request:
# Pull Request 요약

SSE 기반 알림 기능을 구현했습니다. 


## 변경 사항

- 알림 도메인 구현 및 테스트 작성
- 알림 Repository 구현 및 테스트 작성
- 알림 Service 구현 및 테스트 작성
- 알림 Controller 구현 및 테스트 작성
- 알림 Custom annotation 생성 (@NeedNotification)
- 알림 Aspect 구현 (AOP)
- 알림 테이블 DDL 스크립트 작성

## 관련 이슈

- #209 
- #220 

## 개발 유형

- [x] 새로운 기능 개발
- [x] 리팩토링
- [x] 테스트 코드 작성
- [x] 문서 업데이트
- 기타 (아래에 설명 추가)

## 작업 기간

- 작업 시작일: (2024-04-05)
- 작업 종료일: (2024-04-11)

## 유의사항

- @NeedNotification을 메서드에 사용하시면 알림 기능을 이용할 수 있습니다. 다만, 해당 메서드는 NotificationInfo라는 인터페이스 구현체를 반환하도록 하셔야 정상 작동합니다. 

## 참고문헌

- https://develoyummer.tistory.com/112
- https://docs.spring.io/spring-framework/reference/core/aop/introduction-defn.html
- https://product.kyobobook.co.kr/detail/S000001248954

closes #209 #220

JeongUijeong committed:
[feat] 알림 도메인 Entity 생성

JeongUijeong committed:
[refactor] 알림 종류 수정

JeongUijeong committed:
[test] 알림 Entity 테스트 작성

JeongUijeong committed:
[feat] 알림, Emitter 레포지토리 인터페이스 생성

JeongUijeong committed:
[test] Emitter 레포지토리 테스트 작성

JeongUijeong committed:
[feat] Emitter 레포지토리 구현

JeongUijeong committed:
[refactor] 레포지토리 애노테이션 추가

JeongUijeong committed:
[feat] 알림 응답 DTO 생성

JeongUijeong committed:
[feat] SSE 연결 실패 익셉션 생성

JeongUijeong committed:
[test] 알림 서비스 테스트 작성

JeongUijeong committed:
[feat] 알림 서비스 구현

JeongUijeong committed:
[refactor] 알림 타입 별 메시지 추가

JeongUijeong committed:
[feat] 알림이 필요한 메서드에 붙일 커스텀 애노테이션 생성

JeongUijeong committed:
[feat] 알림에 필요한 정보 getter 메서드를 가진 인터페이스 생성

JeongUijeong committed:
[feat] 알림 Aspect 구현

- 포인트컷: @NeedNotification 애노테이션
- 어드바이스: 알림 전송

JeongUijeong committed:
[refactor] 알림 컨트롤러, 서비스 관련 설정  추가

JeongUijeong committed:
[test] 알림 구독 컨트롤러 테스트 작성

JeongUijeong committed:
[feat] 알림 구독 API 컨트롤러 구현

JeongUijeong committed:
[refactor] 읽음 여부 not null 추가

JeongUijeong committed:
[feat] 알림 테이블 ddl 스크립트 작성

chimaek commented on src/main/java/com/fasttime/domain/notification/annotation/NeedNotification.java:
```diff
@@ -0,0 +1,12 @@
+@Target({ElementType.METHOD})
+public @interface NeedNotification {
+
+}
```
코드 리뷰:
1. 알림이 필요한 애노테이션을 나타내는 `NeedNotification` 애노테이션이 설정되었다.
2. 애노테이션을 메소드에 적용할 수 있도록 `@Target({ElementType.METHOD})`가 지정되었다.
3. 실행 시점에 애노테이션이 유지되도록 `@Retention(RetentionPolicy.RUNTIME)`가 지정되었다.
4. 코드에 버그는 없어 보입니다.
5. 개선 제안: 현재는 애노테이션에 파라미터가 없으므로, 필요한 경우 해당 애노테이션에 필요한 정보를 전달할 수 있는 속성을 추가하는 것이 좋습니다.

JeongUijeong replied:
확인했습니다~

chimaek commented on src/main/java/com/fasttime/domain/notification/aop/NotificationAspect.java:
```diff
@@ -0,0 +1,40 @@
+            notificationInfo.getUrl()
+        );
+    }
+}
```
이 코드는 알림 기능 관련 Aspect를 구현한 것으로 보입니다. 여기서 주목해야 할 몇 가지 중요한 점과 개선 사항은 다음과 같습니다:

**잠재적 버그 위험:**
1. `sendNotification` 메소드 내에서 형 변환을 수행하는 부분에서 `result` 객체가 다른 타입인 경우 ClassCastException이 발생할 수 있습니다. 안전한 캐스팅 여부를 확인해야 합니다.
2. `notificationInfo.getNotificationType().getMessage()`에서 NullPointerException이 발생할 수 있습니다. Null 체크를 개선할 필요가 있습니다.

**개선 제안:**
1. `sendNotification` 메소드에 대한 JavaDoc을 추가하여 메소드의 목적과 기능을 명확히 설명합니다.
2. Null 체크를 포함하여 파라미터 및 반환 값의 유효성을 보호하는 코드를 추가합니다.
3. Magic string "`/api/v1/studies/`"을 상수나 환경 변수로 대체하여 유지보수성을 향상시킵니다.
4. 해당 Aspect에 대한 단위 테스트를 작성하여 예상한 동작이 발생하는지 확인합니다.
5. 예외 처리와 로깅을 개선하여 시스템의 안정성을 높입니다.

코드는 깔끔하게 작성되어 있지만 위에 언급된 사항들을 고려하여 더욱 안정적이고 유지보수가 용이한 코드로 개선할 수 있습니다.

JeongUijeong replied:
JavaDoc과 예외 처리 추가했습니다!😊

chimaek commented on src/main/java/com/fasttime/domain/notification/aop/proxy/NotificationInfo.java:
```diff
@@ -0,0 +1,13 @@
+    String getUrl();
+
+    NotificationType getNotificationType();
+}
```
위의 코드 조각에서는 `NotificationInfo` 인터페이스가 정의되어 있습니다. 이 인터페이스는 특정 유형의 알림에 대한 정보를 제공합니다. 

코드 자체에는 버그 또는 잠재적 위험은 보이지 않습니다.
다만, 몇 가지 개선 제안이 있습니다:
1. 메서드 이름에 대해 더 구체적인 설명을 추가하면 좋습니다. 예를 들어 `getUrl()` 메서드가 반환하는 내용이 어디로 연결되는지 세부적으로 명시해주는 것이 도움이 될 수 있습니다.
2. Java 문서 규약을 따라 주석을 추가하여 각 메서드의 역할과 반환 값에 대한 설명을 제공하는 것이 좋습니다.
3. 인터페이스의 용도와 사용 방법을 약간 더 상세히 설명하는 Javadoc을 작성하면 좋습니다.

이러한 조언을 따르면 코드의 가독성과 유지 보수성이 향상될 수 있습니다.

JeongUijeong replied:
NotificationInfo 메서드에 JavaDoc 추가 작성했습니다~😊

chimaek commented on src/main/java/com/fasttime/domain/notification/controller/NotificationController.java:
```diff
@@ -0,0 +1,28 @@
+        @RequestHeader(value = "Last-Event-ID", required = false, defaultValue = "") String lastEventId) {
+        return notificationService.subscribe(securityUtil.getCurrentMemberId(), lastEventId);
+    }
+}
```
이 코드는 NotificationController 클래스를 구현한 것으로 보입니다. 먼저, 이 코드의 주요 기능은 "/api/v2/notifications/subscribe" 엔드포인트에 대해 SSE(Server-Sent Events) 방식으로 클라이언트에게 알림을 제공하는 것으로 파악됩니다.

고려해야 할 사항:
1. parametr lastEventId는 필수가 아닌 선택적(RequestHeader 어노테이션의 required = false)입니다. 이에 대한 설명 없음.
2. SecutiryUtil 클래스가 주입되지만, 어떻게 사용되는지 코드에서 명확히 드러나지 않습니다. 보다 명시적인 사용 방법을 고려하여 개선할 수 있습니다.
3. Subscribe 메서드로 반환된 SseEmitter 객체는 클라이언트 연결을 유지하는데 사용되므로 관리가 중요합니다. 앱 종료 시 자원 해제 및 예외 처리에 대한 계획을 강화해야 합니다.

개선 제안:
1. 주석 추가: 모호한 부분(예: 보안Util 클래스 활용)에 대한 추가 설명이 필요합니다.
2. 세부 정보 표기: return 문에서 반환되는 SseEmitter 객체가 어떻게 관리되는지 등 약간의 더 많은 정보를 추가하여 가독성을 향상시킬 수 있습니다.
3. 예외 처리: 코드에서 발생할 수 있는 예외의 처리 방법을 명확하게 정의하여 예외 처리 전략을 구체화해야 합니다.

전체적으로 나쁘지 않으나 위와 같은 점들을 보강하여 코드를 더욱 안정적이고 이해하기 쉽게 만들 수 있습니다.

JeongUijeong replied:
다음과 같이 수정했습니다! 😊
1. /api/v2/notifications/subscribe API 메서드와 lastEventId에 대한 설명 추가(JavaDoc)
2. SecurityUtil 클래스 사용을 명시
3. SseEmitter 객체의 유효 시간 명시

chimaek commented on src/main/java/com/fasttime/domain/notification/dto/response/NotificationResponseDto.java:
```diff
@@ -0,0 +1,24 @@
+            .url(notification.getUrl())
+            .build();
+    }
+}
```
이 코드는 일반적으로 잘 구조화되어 있으나 개선 사항이 있습니다:

1. `NotificationResponseDto` 클래스 내에서 빌더 패턴(`@Builder`)을 사용하고 있지만, `record` 키워드를 함께 사용하고 있어 혼란 스러울 수 있습니다. 둘 중 하나만 사용하여 일관성 있게 코드를 관리하는 편이 좋습니다.

2. `url` 필드의 유효성(예: URL 형식 확인)을 검증하는 추가적인 코드가 필요할 수 있습니다.

3. `Notification` 객체를 wrapping 할 때 null 체크를 추가하여 NullPointerExceptions 방지를 고려할 수 있습니다.

4. 작은 규모의 코드 변경사항이므로 큰 문제점은 발견되지 않았습니다.

chimaek commented on src/main/java/com/fasttime/domain/notification/entity/Notification.java:
```diff
@@ -0,0 +1,64 @@
+        this.isRead = isRead;
+        this.receiver = receiver;
+    }
+}
```
코드 리뷰:

1. `Notification.java` 클래스는 올바르게 구현되었습니다.
2. `Notification.java` 클래스에 대한 특정 버그 위험이 보이지 않습니다.
3. 몇 가지 개선 제안:
   - `NotificationType` enum에 따라 알림의 유형을 더 자세히 설명하는 주석 추가를 권장합니다.
   - `isRead` 필드 같은 경우 부울(Boolean) 대신 "read" 여부를 나타내는 `boolean isRead()` 와 같은 메소드를 추가하면 코드의 가독성이 좋아질 수 있습니다.

Overall, the code looks well-written and no specific bug risks are visible in the provided information. Consider the suggested improvements for better clarity and maintainability.

JeongUijeong replied:
관련 JavaDoc 추가 했습니다! 😊

chimaek commented on src/main/java/com/fasttime/domain/notification/entity/NotificationType.java:
```diff
@@ -0,0 +1,17 @@
+    NotificationType(String message) {
+        this.message = message;
+    }
+}
```
이 코드 패치는 enum의 기본 구조를 사용하여 NotificationType이라는 알림 유형을 정의하고 있습니다. 리뷰 및 개선 제안은 다음과 같습니다:

1. **개요:**
   - 잘 디자인된 코드 구조입니다. Enum 클래스가 간단하고 명확합니다.
  
2. **잠재적 버그 및 개선 사항:**
   - **Null Pointer 예외**: message 필드가 null 값을 허용하지만, 이는 null일 수 없도록 하는 것이 좋습니다.
   - **I18n & L10n**: 알림 메시지를 외부화하여 국제화 및 지역화를 위한 기능 추가 고려.
   - **Enum 확장성**: 새로운 알림 유형이 더 많아질 경우 대비하여 고려 및 확장 가능한 설계.

3. **개선 제안:**
   - **Builder 패턴 사용**: Enum이 복잡해지면 Builder 패턴 도입을 고려하여 가독성을 향상시킬 수 있습니다.
   - **Switch 문 사용**: Enum 타입에서 switch 문을 사용하여 처리하기 효율적일 수 있습니다.
   - **주석**: 코드에 주석을 추가하여 각 알림 유형의 의미를 명확하게 설명할 수 있습니다.
   - **테스트**: Enum 클래스에 대한 단위 테스트 작성을 고려하여 코드의 안정성 확인.

좀 더 확장 가능한 코드를 위해 Null Pointer 예외 처리와 다국어 지원 등을 검토하는 것이 좋습니다.

JeongUijeong replied:
JavaDoc 추가 및 Null Pointer 예외 처리를 추가 했습니다! 😊

chimaek commented on src/main/java/com/fasttime/domain/notification/exception/SseConnectionFailedException.java:
```diff
@@ -0,0 +1,13 @@
+    public SseConnectionFailedException() {
+        super(ERROR_CODE);
+    }
+}
```
이 코드 조각은 깔끔하고 간단한데 몇 가지 개선 사항이 있습니다:

1. 상수 이름 `ERROR_CODE`를 대문자로 작성하는 것이 Java 표준 관례이므로 유지하는 것이 좋습니다.
2. 주석을 추가하여 클래스 및 메서드의 목적을 설명할 수 있습니다.
3. 필요에 따라 상속 계층에 대해 더 많은 정보가 있으면 좋을 것입니다 - 예를 들어, `ApplicationException`이 어디서 왔는지 등.

잠재적인 버그 위험은 현재 코드에 없어 보입니다. 이 코드는 에러 코드를 사용하여 예외를 생성함으로써 올바르게 구현되었습니다. 코드가 명확하고 가독성이 좋으며 태스크에 적합합니다.

chimaek commented on src/main/java/com/fasttime/domain/notification/repository/EmitterRepository.java:
```diff
@@ -0,0 +1,22 @@
+    void deleteAllByMemberId(long memberId);
+
+    void deleteAllEventCacheByMemberId(long memberId);
+}
```
이 코드에서 발견된 주요 문제 및 개선 제안 사항은 다음과 같습니다:

1. `EmitterRepository` 인터페이스가 명확하고 일관된 메소드명을 사용하고 있으며, 역할이 명확합니다. 그러나 메소드명에 대한 설명 주석을 추가하는 것이 좋습니다.
2. `findAllByMemberId`와 `findAllEventCacheByMemberId` 메소드는 반환 타입이 `Map`이며, 예상보다 복잡할 수 있습니다. 더 구체적인 데이터 구조를 사용하는 것이 이해하기 쉬울 수 있습니다.
3. 메소드명과 파라미터명의 일관성을 유지하면서, 변수명과 메소드명을 더 의미 있는 이름으로 변경할 수 있습니다.
4. 예외 처리에 대한 고려가 필요합니다. 예외 상황에 대한 클래스와 설명이 나와야 합니다.
5. 주석을 통해 각 메소드의 역할과 기능에 대해 자세하게 설명하는 것이 도움이 됩니다.

위 내용들을 고려하여 코드의 가독성과 유지보수성을 향상시킬 수 있습니다.

JeongUijeong replied:
JavaDoc추가 했습니다! 😊

chimaek commented on src/main/java/com/fasttime/domain/notification/repository/EmitterRepositoryImpl.java:
```diff
@@ -0,0 +1,103 @@
+            }
+        });
+    }
+}
```
이 코드는 대체로 안전해 보이지만 두 가지 이슈가 있습니다:

1. `deleteAllByMemberId` 및 `deleteAllEventCacheByMemberId` 메서드에서 반복 중에 Map을 수정하고 있어서 `ConcurrentModificationException`이 발생할 수 있습니다. 이 문제를 해결하기 위해 삭제 연산은 Iterator를 사용하여 진행하는 것이 좋습니다.

2. `deleteAllByMemberId` 와  `deleteAllEventCacheByMemberId` 메서드는 모든 요소를 처리하지 않고 해당 memberId와 관련된 것만을 골라 지우는 점에서 최적화의 여지가 있습니다. 특히, `forEach` 대신 반복문을 활용하여 수정을 줄일 수 있습니다.

이외에는 코드가 명확하고 효율적으로 작성되어 있으며 주요 기능들이 구현되어 있습니다.

chimaek commented on src/main/java/com/fasttime/domain/notification/repository/NotificationRepository.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface NotificationRepository extends JpaRepository<Notification, Long> {
+
+}
```
주어진 코드 조각은 Spring Data JPA를 사용해 Notification 엔티티에 대한 Repository를 정의하고 있습니다.

개선 제안:
1. 주석: 코드 상단에 해당 리포지토리의 목적이나 특정 사항을 설명하는 주석을 추가하는 것이 좋습니다.
2. 비즈니스 로직: 현재는 단순히 JpaRepository를 확장하고 있으므로, 해당 리포지토리에 추가 비즈니스 로직을 구현할 수 있도록 고려해야 합니다.
3. 테스트: CRUD(Create, Read, Update, Delete) 작업에 대한 테스트 케이스 작성을 고려해야 합니다.
4. 예외 처리: 예외 상황 발생 시 적절한 예외 처리를 고려하여 기능을 강화할 수 있습니다. 

버그 위험 :
1. 오타: 패키지 이름, 클래스 이름 또는 메서드 이름 등에 오타가 있는지 확인해야 합니다.
2. 설정 오류: 스프링의 구성 파일이나 의존성 주입에 문제가 있을 수 있으니 관련 설정을 확인해야 합니다.

이 코드는 기본적으로 올바르게 보이며, 추가적인 개선 사항이나 버그는 없어 보입니다.

chimaek commented on src/main/java/com/fasttime/domain/notification/service/NotificationService.java:
```diff
@@ -0,0 +1,17 @@
+        String content,
+        String url
+    );
+}
```
주요 문제점:
1. `NotificationService` 인터페이스는 `subscribe`와 `send` 메서드를 갖고 있습니다. 이는 알림 서비스 전반을 다루지만, 여기서 서브스크립션과 메시지 전송만 다루도록 분리하는 것이 좋을 수 있습니다.
2. 주석 없이 메서드의 목적을 설명하는 javadoc이 생성되어야 합니다.
3. `send` 메서드 내에서 parameter validation이 필요할 것으로 보입니다.

개선 제안:
1. `NotificationService` 인터페이스를 `SubscriptionService`와 `MessageService`로 나누는 것을 고려해보세요.
2. 메서드에 대한 자세한 설명을 추가하고, 파라미터 유효성 검사(validation)를 구현하세요.
3. `NotificationType` 대신 enum 사용을 고려해보세요.
4. 상황에 따라 에러 처리 및 로깅을 추가하세요.

chimaek commented on src/main/java/com/fasttime/domain/notification/service/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,151 @@
+            sendNotification(emitter, eventId, emitterId, NotificationResponseDto.of(notification));
+        });
+    }
+}
```
이 코드는 SSE(Server-Sent Events) 알림 서비스를 구현한 것으로 보입니다. 여러 개선 및 버그 위험 요소들이 있습니다:

1. **개선 사항:**
    - `makeId()`에서 현재 시간을 포함하여 Emitter 식별자를 만드는 방식은 충돌 가능성이 있습니다.
    - Exception 처리가 세밀하지 않습니다. 예외 처리를 더 구체적으로 나누어 개선할 수 있습니다.
    - `sendLostData()` 메서드 내부의 Map 반복문 처리는 복잡합니다. 더 효율적인 방법 고려 필요.
    - 주석이 자세하고 이해하기 쉽지만, 가독성을 높이기 위해 비즈니스 로직에 대한 주석을 추가하는 것이 좋습니다.

2. **버그 위험 요소:**
    - `hasLostData()` 메서드는 `lastEventId`가 비어있을 때 `true`를 반환합니다. 의도와 다르게 동작할 수 있음.
    - `emitterRepository.saveEventCache(emitterId, notification);` 에서 저장된 Cache 관리 및 정리가 필요.

3. **보안 고려 사항:**
    - 안전하지 않은 작업을 해결하기 위해 필수적으로 필요한 권한 확인 및 예외 처리가 누락되어 있습니다.
  
개발 및 유지보수를 용이하게 하기 위해 코드 리팩터링 및 예외처리 기능을 개선하는 것이 추천됩니다.

chimaek commented on src/main/java/com/fasttime/global/exception/ErrorCode.java:
```diff
@@ -56,6 +56,9 @@ public enum ErrorCode {
+
     // 5xx
     INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버 내부 에러");
 
```
코드 리뷰:

1. SSE_CONNECTION_FAILED 에러 항목이 추가되어서 올바르게 섹션이 정의되었습니다. 
2. 코드 변경은 ErrorCodesEnum 클래스에서 이루어지며, 추가된 새로운 상수는 서버 내부 오류를 나타냅니다.
3. 주석은 간략하면서도 목적을 잘 설명하고 있습니다.
4. 코딩 스타일 및 형식에 대한 일관성 유지를 위해 코드베이스 내 다른 부분과 일관된 형태로 작성되었는지 확인하세요.

추가적으로 향후 유지보수를 고려하여 테스트 케이스가 적용되었는지 확인하고, 예외 처리 메커니즘이 적절히 설정되었는지 검토해 보시기 바랍니다.

chimaek commented on src/main/resources/db/migration/V23__Add_notification_table.sql:
```diff
@@ -0,0 +1,16 @@
+);
+
+ALTER TABLE notification
+    ADD CONSTRAINT FK_NOTIFICATION_ON_MEMBER FOREIGN KEY (member_id) REFERENCES member (id);
```
이 코드 패치의 몇 가지 개선 및 버그 위험 사항:

1. ENUM 타입을 사용하면 새로운 유형을 추가/제거하기가 까다로울 수 있으며, 데이터 일관성이 유지되지 않을 위험이 있습니다. 대신 이를 별도의 테이블과 외래 키를 활용하여 구현하는 것이 더 좋을 수 있습니다.
2. is_read 필드의 데이터 형식으로 BIT(1) 보다 BOOLEAN을 사용하는 것이 더 명확하고 표준적입니다.
3. 생성일(created_at), 업데이트일(updated_at), 삭제일(deleted_at)에는 기본값(DEFAULT CURRENT_TIMESTAMP)을 설정하여 자동으로 값이 입력되도록 구현하는 것이 좋습니다.
4. 외래 키 제약 조건을 사용할 때 참조 무결성을 해치지 않게 신중하게 선택해야 합니다. 올바른 회원 ID만 삽입되도록 확인이 필요합니다.

개정된 코드 예시:

```sql
CREATE TABLE notification
(
    id         BIGINT AUTO_INCREMENT NOT NULL,
    type_id    BIGINT NOT NULL,
    content    VARCHAR(255) NULL,
    url        VARCHAR(255) NULL,
    is_read    BOOLEAN NOT NULL DEFAULT false,
    member_id  BIGINT NULL,
    created_at datetime DEFAULT CURRENT_TIMESTAMP,
    updated_at datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
    deleted_at datetime NULL,
    CONSTRAINT pk_notification PRIMARY KEY (id),
    CONSTRAINT fk_notification_type FOREIGN KEY (type_id) REFERENCES notification_types(id),
    CONSTRAINT fk_notification_member FOREIGN KEY (member_id) REFERENCES member (id)
);
```

위의 수정사항은 코드의 가독성 및 데이터 일관성을 향상시키며, 버그 리스크를 줄일 수 있습니다.

chimaek commented on src/test/java/com/fasttime/domain/notification/unit/controller/NotificationControllerTest.java:
```diff
@@ -0,0 +1,44 @@
+            verify(notificationService, times(1)).subscribe(any(long.class), any(String.class));
+        }
+    }
+}
```
주요 문제나 개선 사항은 다음과 같습니다:

1. 주석이 부족합니다. 코드의 각 부분에 대한 설명이 더 필요합니다.
2. 테스트 케이스가 하나만 있습니다. 더 많은 시나리오 및 에지 케이스에 대한 테스트를 추가하는 것이 좋습니다.
3. 테스트 메서드 이름이 조금 더 명확하고 구체적으로 작성될 수 있습니다.
4. `notificationService`의 선언이 없습니다. 이것은 외부에서 어디서 주입되는지 파악할 수 없습니다.
5. `@DisplayName` annotation 을 통해 한글로 명시적으로 테스트 케이스에 대한 내용을 기술해주어 가독성 및 이해도를 높일 수 있습니다.

추가로, 테스트 커버리지를 높이고 `notificationService`에 대한 목(mock)을 사용하여 더 복잡한 시나리오를 다루는 테스트들을 추가하는 것이 좋습니다.

chimaek commented on src/test/java/com/fasttime/domain/notification/unit/entity/NotificationTest.java:
```diff
@@ -0,0 +1,41 @@
+            .build()
+        );
+    }
+}
```
코드 리뷰:

1. 코드에서 큰 버그는 없어 보입니다.
2. 단위 테스트가 적절하게 구성되어 있습니다.
3. 개선 제안:
   - Member 객체를 직접 생성하는 대신, Mockito 또는 다른 mocking 프레임워크를 사용하여 가짜 Member 객체를 만들 수 있습니다.
   - 각 단계(주어진 것, 실행, 결과)에 대한 주석을 추가하여 코드의 의도를 명확히 할 수 있습니다.
   - 더 많은 테스트 케이스를 추가하여 코드 커버리지를 향상시킬 수 있습니다.

이 코드 조각에서는 주요 버그나 문제가 발견되지 않았으며, 단위 테스트가 올바르게 작성되었습니다. 위의 개선 제안을 고려하여 코드의 가독성과 유지보수성을 향상시킬 수 있습니다.

chimaek commented on src/test/java/com/fasttime/domain/notification/unit/repository/EmitterRepositoryImplTest.java:
```diff
@@ -0,0 +1,212 @@
+        // then
+        Assertions.assertEquals(emitterRepository.findAllEventCacheByMemberId(memberId).size(), 0);
+    }
+}
```
코드 리뷰:
1. `DEFAULT_TIMEOUT` 상수는 좋지만, 60L * 1000L * 60L 대신에 밀리세컨드 단위로 표현하거나 주석으로 단위를 명시하는 것이 더 명확할 수 있습니다.
2. 많은 메서드에서 시간 기반의 대기 (`Thread.sleep`)를 사용하고 있습니다. 이것은 테스트 안정성과 실행 속도에 영향을 줄 수 있으므로 해당 부분을 최소화하는 것이 좋습니다.
3. 테스트 이름이 명확하게 기능을 설명하지만, 코드 커버리지가 충분한지 확인이 필요합니다.
4. 지역 변수명이나 메서드명을 더 명확하게 하여 가독성을 향상할 수 있습니다.
5. 테스트 케이스 간의 의존성을 완전히 없애는 게 좋을 수 있습니다.

전반적으로 테스트 케이스들은 목적에 맞게 잘 작성되어 있습니다. 코드의 가독성, 성능 및 안정성에 대한 개선 사항을 고려해보세요.

chimaek commented on src/test/java/com/fasttime/domain/notification/unit/service/NotificationServiceImplTest.java:
```diff
@@ -0,0 +1,102 @@
+                "www.boocam.net/studies/1"
+            ));
+    }
+}
```
이 코드 조각은 대체로 깔끔하며 안전해 보입니다. 그러나 몇 가지 개선 제안이 있습니다:

1. `subscribeWillSuccess` 및 `sendWillSuccess` 테스트 메소드: 각 테스트는 독립적이고 동일한 설정을 반복합니다. 공통 설정을 `@BeforeEach` 애노테이션으로 이동하여 중복을 줄일 수 있습니다.

2. 주석: 메소드 상단에 "given", "when then"과 같은 주석을 추가하여 테스트의 구조를 명확히 할 수 있습니다.

3. 테스트 케이스 더 세분화: `subscribe` 및 `send` 테스트가 서로 다른 시나리오에 대한 단일 메소드로 결합되었습니다. 더 작은 테스트 케이스로 나누면 테스트 실패 시 디버깅 및 유지 관리가 더 쉬울 수 있습니다.

4. 매직 넘버: `DEFAULT_TIMEOUT`과 같은 매직 넘버는 설명이 필요합니다. 해당 값이 어떤 시간을 나타내는지 주석으로 설명하는 것이 좋습니다.

5. `emitterId` 생성: 현재는 `System.currentTimeMillis()` 함수를 사용하여 `emitterId`를 생성하는데, 이는 유닉스 시간을 기반으로 하기 때문에 멀티 스레딩 환경에서 충돌 위험이 있습니다. 이를 해결하기 위해 좀 더 안전한 방식으로 고유한 ID를 생성하는 방법을 고려할 수 있습니다.

이외에는 기능적인 측면에서 큰 문제가 보이지 않습니다.

chimaek commented on src/test/java/com/fasttime/util/ControllerUnitTestSupporter.java:
```diff
@@ -76,4 +78,7 @@ public abstract class ControllerUnitTestSupporter {
+
+    @MockBean
+    protected NotificationService notificationService;
 }
```
코드 리뷰:
1. 새로운 `NotificationController` 및 `NotificationService`가 추가되었습니다. 이들은 문제 없이 추가된 것으로 보입니다.
2. `@WebMvcTest` 어노테이션에서 `NotificationController.class`가 제대로 포함되었으나, 오타가 존재합니다. "contoller" 대신에 "controller"로 수정해야 합니다.
3. 테스트 유틸리티 클래스인 `ControllerUnitTestSupporter`에 `NotificationService`에 대한 `MockBean`이 적절히 추가되었습니다. 필수적인 변경 사항으로 보입니다.
4. 코드를 더 읽지 않았기 때문에 완전한 리뷰는 제공할 수 없습니다. 그러나 주요 변경 사항들은 깔끔하게 반영된 것으로 보이며, 개발자의 기여에 따라 추가적인 조치가 필요할 수 있습니다.

JeongUijeong replied:
확인해봤는데 오타를 못 찾았습니다...😅

JeongUijeong committed:
[refactor] 메서드에 JavaDoc을 추가해 목적과 기능 명시

JeongUijeong committed:
[refactor] 알림 전송 메서드에서 다운캐스팅 에러 처리 추가

JeongUijeong committed:
[refactor] NotificationInfo 인터페이스에 JavaDoc 추가

JeongUijeong committed:
[refactor] 알림 컨트롤러 코드리뷰 반영

- /api/v2/notifications/subscribe API 메서드와 lastEventId에 대한 설명 추가(JavaDoc)
- SecurityUtil 클래스 사용을 명시
- SseEmitter 객체의 유효 시간 명시

JeongUijeong committed:
[refactor] 알림 Entity JavaDoc 작성 및 알림 종류 enum에 대한 주석 추가

JeongUijeong committed:
[refactor] 알림 유형 열거형에 JavaDoc 수정 및 알림 메시지 NotNull 처리 추가

JeongUijeong committed:
[refactor] Emitter 레포지토리 인터페이스에 JavaDoc 작성

ghrltjdtprbs reviewed (approved):
와우...👍

JeongUijeong committed:
[fix] ErrorCode 컨플릭트 해결

chimaek commented on src/main/java/com/fasttime/global/exception/ErrorCode.java:
```diff
@@ -61,6 +61,11 @@ public enum ErrorCode {
+
     // 5xx
     INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버 내부 에러");
 
```
위 코드 조각의 주요 문제점은 새로 추가된 에러 코드 중 일부가 동일한 HTTP 상태 코드인 HttpStatus.INTERNAL_SERVER_ERROR를 공유한다는 점입니다. 이렇게 하면 나중에 수정이 필요하거나 변경할 경우 혼란스러울 수 있습니다.

향후 유지보수성을 고려하여 다음의 개선 제안을 할 수 있습니다:
1. 각 에러 코드에 대해 고유한 HTTP 상태 코드 사용을 고려하여 중복을 피하십시오.
2. 에러 메시지를 더 명확하게 작성하여 디버깅 및 유지 관리를 용이하게 합니다.
3. 에러 코드 설명이 현재 언어로 한국어로 번역되어 있는데, 일반적으로 영어로 유지하는 것이 권장됩니다.

이러한 수정 사항은 코드의 가독성을 향상시키고 잠재적인 버그 위험을 줄일 수 있습니다.

JeongUijeong committed:
[fix] 컨플릭트 해결

chimaek commented on src/main/java/com/fasttime/global/exception/ErrorCode.java:
```diff
@@ -66,6 +66,10 @@ public enum ErrorCode {
+        "알림 전송 기능을 사용할 메서드의 반환값이 알림 전송을 지원하지 않는 자료형입니다."),
 
     // 5xx
     INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버 내부 에러");
```
코드 리뷰:

1. 새로운 에러 코드가 추가되었지만, 주석이 일부 누락되어 있습니다. 에러 코드에 대한 목적과 사용 사례가 좀 더 명확히 설명돼야 합니다.
2. "SSE_CONNECTION_FAILED"와 "UNSUPPORTED_DATA_TYPE_FOR_NOTIFICATION_SENDING"의 상태 코드가 "INTERNAL_SERVER_ERROR"로 동일합니다. 이것이 어떤 종류의 오류인지 명시해야 할 것입니다.
3. Enum 이름이 "ErrorCode"이지만, 정확히는 HTTP 응답 코드와 연관된 메시지를 가지고 있습니다. 이름을 더 명확하게 변경하는 것이 좋을 수 있습니다.

개선 제안:
1. 모든 에러 코드에 대한 상세 설명을 주석으로 추가하여 코드 사용자가 명확하게 알 수 있도록 합니다.
2. Enum 이름을 보다 명확한 이름으로 변경하여 코드 의도를 더 잘 전달하도록 합니다.
3. 상태 코드("HttpStatus")에 대한 표준화된 포맷 및 상수 사용을 고려하면 유지보수와 코드 가독성 측면에서 도움이 될 수 있습니다.
